File: examples/assume-contracts/src/main/java/example/env/Buckets.java
package example.env;

/**
 * The "environment as a fixed value" case (the okio {@code HASH_BUCKET_COUNT} shape). A static
 * {@code <clinit>} reads a deterministic, unanalyzable query - {@link Environment#bucketCount()} on the
 * process-wide {@link #ENV} - into a {@code static final} bound that sizes an array and a fill loop.
 *
 * <p>Without help that bound is SYMBOLIC (the call has no analyzed body, so JBMC nondet-stubs it), so
 * {@code count()} can be anything and the array bound can't be closed.
 * {@code Bmc.assumeStable(ENV::bucketCount, n -> n == 8)} pins it to one fixed value for the whole run -
 * including this {@code <clinit>}, a call site a local {@code assume} can't reach - so the bound is the
 * concrete 8 and the proof verifies.
 */
public final class Buckets {

    /** The process-wide environment. Held in a static so the {@code <clinit>} below reads it; the proof
     *  pins {@code ENV.bucketCount()} via {@code assumeStable}. */
    public static Environment ENV;

    /** Read once in {@code <clinit>} from the (unanalyzable) environment query. */
    private static final int COUNT = ENV.bucketCount();

    private final int[] slots = new int[COUNT];

    public Buckets() {
        for (int i = 0; i < COUNT; i++) {
            slots[i] = i;
        }
    }

    /** The bound the {@code <clinit>} captured. */
    public int count() {
        return COUNT;
    }

    /**
     * The last filled slot's value, {@code slots[COUNT - 1]} (require {@code COUNT >= 1}). By
     * construction the fill loop sets it to {@code COUNT - 1} - proving that requires the loop fully
     * unwound, which needs the bound pinned to a fixed value.
     */
    public int lastSlot() {
        return slots[COUNT - 1];
    }
}

File: examples/assume-contracts-kotlin/src/main/kotlin/example/env/Buckets.kt
package example.env

/**
 * The "environment as a fixed value" case (the okio HASH_BUCKET_COUNT shape). A static initializer
 * reads a deterministic, unanalyzable query - [Environment.bucketCount] on the process-wide [ENV] -
 * into a final bound that sizes an array and a fill loop.
 *
 * Without help that bound is SYMBOLIC (the call has no analyzed body, so JBMC nondet-stubs it), so
 * count() can be anything and the array bound can't be closed.
 * `Bmc.assumeStable(env::bucketCount) { it == 8 }` pins it to one fixed value for the whole run -
 * including this static initializer, a call site a local assume can't reach - so the bound is the
 * concrete 8 and the proof verifies.
 */
class Buckets {

    private val slots = IntArray(COUNT)

    init {
        for (i in 0 until COUNT) {
            slots[i] = i
        }
    }

    /** The bound the static initializer captured. */
    fun count(): Int = COUNT

    /**
     * The last filled slot's value, slots[COUNT - 1] (require COUNT >= 1). By construction the fill
     * loop sets it to COUNT - 1 - proving that requires the loop fully unwound, which needs the bound
     * pinned to a fixed value.
     */
    fun lastSlot(): Int = slots[COUNT - 1]

    companion object {
        /** The process-wide environment. Non-null (a benign default the proof overrides) so the COUNT
         *  initializer below reads it with a plain interface call and NO Kotlin null-check - the proof
         *  pins ENV.bucketCount() via assumeStable, which redirects that call regardless of the receiver. */
        @JvmField
        var ENV: Environment = DefaultEnvironment

        /** Read once in the static initializer from the (unanalyzable) environment query. */
        private val COUNT: Int = ENV.bucketCount()
    }
}
